cmd/iptables-wrapper.go

```
@@ -0,0 +1,174 @@
+the command and the arguments to it.
+*/
+func main() {
+	if err := checkIptablesNFTVersion(); err != nil {
+		fmt.Fprintf(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	iptablesMode, err := getIptablesMode()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	var iptablesBinary string
+
+	switch iptablesMode {
+	case iptablesNFTMode:
+		iptablesBinary = iptablesNFTPath
+	case iptablesLegacyMode:
+		iptablesBinary = iptablesLegacyPath
+	default:
+		fmt.Fprintf(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	args := make([]string, 0)
+	args = append(args, filepath.Base(os.Args[0]))
+	if len(os.Args) > 0 {
+		args = append(args, os.Args...)
+	}
+
+	// Buffers are faster than standard string reading
+	var output bytes.Buffer
+
+	cmdIptables := &exec.Cmd{
+		Path:   iptablesBinary,
+		Args:   args,
+		Stdout: &output,
+	}
+
+	if err := cmdIptables.Run(); err != nil {
+		fmt.Fprintf(os.Stderr, err.Error())
+		os.Exit(1)
+	}
+
+	fmt.Print(output.String())
+
+}
+
+func checkIptablesNFTVersion() error {
+	outNft, err := exec.Command(iptablesNFTPath, "iptables", "--version").Output()
+	if err != nil {
+		return err
+	}
+	if regexIPTVers.Match(outNft) {
+		return errors.New("ERROR: iptables 1.8.0 - 1.8.3 have compatibility bugs. Upgrade to v1.8.4 or newer.")
+	}
+	return nil
+}
+
+// getIptablesMode uses the tries to find the pre-existing "KUBE-IPTABLES-HINT"
+// or "KUBE-KUBELET-CANARY" to determine which iptables mode should be used
+func getIptablesMode() (string, error) {
+	/*
+		In kubernetes 1.17 and later, kubelet will have created at least
+		one chain in the "mangle" table (either "KUBE-IPTABLES-HINT" or
+		"KUBE-KUBELET-CANARY"), so check that first, against
+		iptables-nft, because we can check that more efficiently and
+		it's more common these days.
+	*/
+
+	outNft, err := exec.Command(iptablesNFTPath, "iptables-save",
+		"-t", "mangle").Output()
+
+	if err != nil {
+		return "", err
+	}
```

the shell version just discards errors